### DISCUSSION TITLE
How to specify/get prompt for additional parameters?

### DISCUSSION THREAD

**Discussion by pamelafox** [Azure/azure-dev#906] [Q&A]
Hi! I'm trying to use azd for a Bicep that has several additional required parameters, and the deploy fails if they aren't set. However, I can't figure out how to get azd to prompt me to set those parameters, like it prompts for name and location.

Here's my parameters file:

```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
      "name": {
        "value": "${AZURE_ENV_NAME}"
      },
      "location": {
        "value": "${AZURE_LOCATION}"
      },
      "homeIP": {
        "value": "${HOME_IP_ADDRESS}"
      },
      "databasePassword": {
        "value": "${DATABASE_PASSWORD}"
      }
    }
  }
```

Corresponding to this bicep:

```
targetScope = 'subscription'

@minLength(1)
@maxLength(64)
@description('Name to prefix all resources')
param name string = 'babybuddy'

@minLength(1)
@description('Primary location for all resources')
param location string = 'eastus'

@description('Home IP address')
param homeIP string

@secure()
param databasePassword string
```

How could I get it to prompt for homeIP and databasePassword? Or, is there a way to pass them into the `azd up` command?

Thanks! Sorry if I'm just completely missing something here.

**Comment by charris-msft** [comment]
@puicchan - do you know if this is possible today?

If/when we can do it at the CLI, I'd like to bridge that across to the Deploy to Azure button work so the same params get exposed in the portal.

**Comment by puicchan** [comment]
I set the values in https://github.com/puicchan/app-templates-integration-services/blob/main/infra/main.parameters.json

@jongio - is there another way to get `azd up` to prompt for the values? 

{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
      "environmentName": {
        "value": "${AZURE_ENV_NAME}"
      },
      "location": {
        "value": "${AZURE_LOCATION}"
      },
      "publisherEmail": {
        "value": "xxx@microsoft.com"
      },
      "publisherName": {
        "value": "PC"
      }      
    }
  }

**Comment by pamelafox** [comment]
@puicchan Did that work for you, setting in main.parameters.json? I thought it didn't when I tried it originally. I'll try it again.

**Comment by puicchan** [comment]
Yes it did.

**Comment by pamelafox** [comment]
Hm, I just tried `azd up` on your repo, and the azd prompt didn't ask me for those additional parameters:

<img width="1205" alt="Screenshot 2022-11-17 at 8 19 59 PM" src="https://user-images.githubusercontent.com/297042/202616016-28c3f2b5-f204-4e28-ba34-6bb59fbb37b0.png">

I was hoping it would ask for them the same way it asks for environment/location. Or is the idea that developers have to fill out `main.parameters.json` before running `azd up`?

**Comment by puicchan** [comment]
Yes, I set the values in `main.parameters.json`. That's why I asked @jongio if there is a way to get azd to prompt for input.

**Reply by pamelafox** [reply]
Ahh okay! I will do that for now and make sure to mention in the readme which parameters need to be updated.

**Comment by vhvb1989** [comment]
@pamelafox , azd is expected to prompt for the value.  Using `main.parameters.json` is just a way to map the parameter to some value from the environment.
I reviewed the implementation and there's currently a bug because azd is hiding the prompt behind the spinner:

![image](https://user-images.githubusercontent.com/24213737/202788727-f7101cfe-de59-4aff-a37e-e134fa215fac.png)

This also makes `Compiling Bicep template` to never ends, because it stays waiting for a value.

You can work around this (while the issue is fixed) by using `-o json` when running `azd provision`.  This will disable the progress spinner and let you see the prompt:

![image](https://user-images.githubusercontent.com/24213737/202789260-07e838e1-26b6-4638-aff5-06b85e98c026.png)

You can also continue using `main.parameters.json` to map the param to some value within your environment (which might be usually better as it would let you set the value in your env (or per azd environment) instead of getting the prompt.  However, if the value is not set in the environment, we should get the prompt.

**Reply by pamelafox** [reply]
Ooh, fascinating. Thanks for replying and filing issue #1191
